starter/src/main/java/com/example/starter/base/impl/MainView.java

```diff
@@ -15,15 +19,29 @@
 @Route(value = "", layout = MainLayout.class)
 public class MainView extends VerticalLayout {
 
+    public GreetService getGreetService() throws Exception {
```

Comment:
This makes the main project effectively depends on  `GreetService`  impl which exists only in the different bubdle.
So it will fail if the module is not deployed.

The starter project should work without any additional bundle deployed.

So either custom default GreetService should be instantiated right here if it's not available or 
default  GreetService  impl should be declared in the starter project with minimal rank .
The impl in the separate bundle should use either default rank (0) or the maximum rank so it takes a precedence.

That way the app always will work without any extra module without failures but once the greetservice is deployed the behavior will change.

Reply:
The main project (starter) depends only on the interface, which in the starter bundle, not the actual implementation which is in optional greetservice bundle. It deploys as before without the optional greetservice bundle

Should probable rename the greetservice bundle (greetservice-impl or similar) to make it more obvious.

Reply:
Now it is greetservice-impl and the top level readme underlines that it is also optional bundle. I was pondering whether to introduce an optional bundle just for the GreetService interface (greetservice-api), but finally though that maybe it is not necessary. After all, we are not trying to teach OSGi people how to do things, just trying to present that things work fine with Vaadin UIs. But I'm fine adding greetservice-api as well if you think it would make the example clearer.